### PR TITLE
OpenFermion estimates using ERI

### PR DESCRIPTION
**Pull request by max-radin**
## Description

Currently, the benchq integration with OpenFermion performs resource estimates on pyscf meanfield objects. However, we have several cases where we'd like to get OpenFermion resource estimates using only the molecular integrals. (For example, the HRL anti-corrosive instances and the L3-Zapata catalysis instances.)

This pull request refactors the OpenFermion integration to accept molecular integrals rather than a pyscf meanfield object.

For users who still wish to get resource estimates for pyscf objects, the one- and two-body integrals can be obtained as follows:
```python
from benchq.resource_estimation import get_single_factorized_qpe_resource_estimate
from openfermion.resource_estimates.molecule import pyscf_to_cas

h1, eri_full, _, _, _ = pyscf_to_cas(mean_field_object)
resource_estimate = get_single_factorized_qpe_resource_estimate(h1, eri_full, rank)
```

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [x] I have updated documentation.

### CODE REVIEW

**Review by mkodrycka — APPROVED** [review]

File: src/benchq/resource_estimation/_compute_lambda_sf.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import numpy as np
+
+
+def compute_lambda(h1, eri_full, sf_factors):
+    """Compute lambda for Hamiltonian using SF method of Berry, et al.
```
[the remark this answers is not in the document]
Comment (aakunitsa) [reply src/benchq/resource_estimation/_compute_lambda_sf.py:21]: I think that's the one Max is referring to: https://arxiv.org/pdf/1902.02134.pdf

File: src/benchq/resource_estimation/openfermion_re.py [hunk 2]
```diff
@@ -1,69 +1,92 @@
...
-def model_toffoli_and_qubit_cost_from_single_factorized_mean_field_object(
-    mean_field_object, rank, DE, CHI
-):
-    num_orb = len(mean_field_object.mo_coeff)
+
+def get_single_factorized_qpe_toffoli_and_qubit_cost(
+    h1: np.ndarray,
+    eri: np.ndarray,
+    rank: int,
+    allowable_phase_estimation_error: float,
```
Comment (aakunitsa) [inline src/benchq/resource_estimation/openfermion_re.py:19]: Are the reasonable defaults for some of the parameters in this function? For example, `allowable_phase_estimation_error`?

File: src/benchq/resource_estimation/openfermion_re.py [hunk 3]
```diff
@@ -1,69 +1,92 @@
...
-    if not meanfield_object._eri.shape == (meanfield_object._eri.shape[0],) * 4:
-        meanfield_object._eri = _get_unsymmetrized_eri(meanfield_object)
+    Args:
+        h1 (np.ndarray): Matrix elements of the one-body operator that includes kinetic
+            energy operator and electorn-nuclear Coulomb operator.
+        eri (np.ndarray): Four-dimensional array containing electron-repulsion
+            integrals.
+        rank (int): Rank of the factorization.
+        allowable_phase_estimation_error (float): Allowable error in phase estimation.
+            Corresponds to epsilon_QPE in the paper.
```
[the remark this answers is not in the document]
Comment (aakunitsa) [reply src/benchq/resource_estimation/openfermion_re.py:74]: PRX Quantum 2, 030305 - it is mentioned at the beginning of the docstring!

File: src/benchq/resource_estimation/openfermion_re.py [hunk 4]
```diff
@@ -1,69 +1,92 @@
...
+        lam,
+        allowable_phase_estimation_error,
+        L=rank,
+        chi=bits_precision_state_prep,
+        stps=stps1,
     )
     return sf_total_toffoli_cost, sf_logical_qubits
 
 
-def _get_unsymmetrized_eri(meanfield_object: pyscf.scf.hf.SCF) -> np.ndarray:
```
Comment (aakunitsa) [inline src/benchq/resource_estimation/openfermion_re.py:33]: I think it makes sense to keep this as a utility function - would you agree @max-radin?

**Review by aakunitsa — CHANGES_REQUESTED** [review]
Thank you for preparing the PR, Max! I think it is mostly ready to be merged, but I have a few minor requests & questions.

File: src/benchq/resource_estimation/_compute_lambda_sf.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mkodrycka) [reply src/benchq/resource_estimation/_compute_lambda_sf.py:21]: @aakunitsa Thanks!

**Review by aakunitsa — APPROVED** [review]

**Pull request merged**